crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2466,16 +2480,138 @@ impl<'db> TypeInferenceBuilder<'db> {
+                                    right_ty.display(self.db)
+                                ),
+                            );
+                            match op {
+                                // `in, not in, is, is not` always return bool instances
+                                ast::CmpOp::In
+                                | ast::CmpOp::NotIn
+                                | ast::CmpOp::Is
+                                | ast::CmpOp::IsNot => {
+                                    builtins_symbol_ty(self.db, "bool").to_instance(self.db)
+                                }
+                                // Other operators can return arbitrary types
+                                _ => Type::Unknown,
+                            }
+                        })
+                }),
+            ops.len(),
+        )
+    }
+
+    /// Infers the type of a binary comparison (e.g. 'left == right'). See
+    /// `infer_compare_expression` for the higher level logic dealing with multi-comparison
+    /// expressions.
+    ///
+    /// If the operation is not supported, return None (we need upstream context to emit a
+    /// diagnostic).
+    fn infer_binary_type_comparison(
+        &mut self,
+        left: Type<'db>,
+        op: ast::CmpOp,
+        right: Type<'db>,
+    ) -> Option<Type<'db>> {
+        // Note: identity (is, is not) is unreliable in Python and not part of the language specs.
+        // - `[ast::CompOp::Is]`: return `false` if different, `bool` if the same
+        // - `[ast::CompOp::IsNot]`: return `true` if different, `bool` if the same
+        match (left, right) {
+            (Type::IntLiteral(n), Type::IntLiteral(m)) => match op {
+                ast::CmpOp::Eq => Some(Type::BooleanLiteral(n == m)),
+                ast::CmpOp::NotEq => Some(Type::BooleanLiteral(n != m)),
+                ast::CmpOp::Lt => Some(Type::BooleanLiteral(n < m)),
+                ast::CmpOp::LtE => Some(Type::BooleanLiteral(n <= m)),
+                ast::CmpOp::Gt => Some(Type::BooleanLiteral(n > m)),
+                ast::CmpOp::GtE => Some(Type::BooleanLiteral(n >= m)),
+                ast::CmpOp::Is => {
+                    if n == m {
+                        Some(builtins_symbol_ty(self.db, "bool").to_instance(self.db))
+                    } else {
+                        Some(Type::BooleanLiteral(false))
+                    }
+                }
+                ast::CmpOp::IsNot => {
+                    if n == m {
+                        Some(builtins_symbol_ty(self.db, "bool").to_instance(self.db))
+                    } else {
+                        Some(Type::BooleanLiteral(true))
+                    }
+                }
+                // Undefined for (int, int)
+                ast::CmpOp::In | ast::CmpOp::NotIn => None,
+            },
+            (Type::IntLiteral(_), Type::Instance(_)) => self.infer_binary_type_comparison(
+                builtins_symbol_ty(self.db, "int").to_instance(self.db),
```

Comment:
we added a helper method for this recently:

https://github.com/astral-sh/ruff/blob/cc1f766622bd27c24e47362503f44f8545710c6f/crates/red_knot_python_semantic/src/types.rs#L392-L394

```suggestion
                Type::builtin_int_instance(self.db),
```

Reply:
You could consider adding a `Type::builtin_bool_instance` method as part of this PR as well, if you like. But definitely not mandatory :-)

Reply:
Interesting... I agree that the current situation isn't _great_, but I don't know that an enum is definitely the way to go. There's a lot of builtins; it might be a pretty big enum if you want to have them all there! And some builtins we are just going to need to access a lot more than others, so it does make sense to treat different builtins differently, I think. But anyway, definitely curious to see what you come up with :-)